resource/wrapper/retryresource/crud_resource_test.go

```diff
@@ -1,364 +1,446 @@
+			},
+		},
+		{
+			ErrorCount:  2,
+			ErrorMethod: "GetCurrentState",
+			ExpectedMethodOrder: []string{
+				"GetCurrentState",
+				"GetCurrentState",
+				"GetCurrentState",
+				"GetDesiredState",
+				"NewUpdatePatch",
+				"ApplyCreatePatch",
+				"ApplyDeletePatch",
+				"ApplyUpdatePatch",
+			},
+		},
+		{
+			ErrorCount:  2,
+			ErrorMethod: "ApplyUpdatePatch",
+			ExpectedMethodOrder: []string{
+				"GetCurrentState",
+				"GetDesiredState",
+				"NewUpdatePatch",
+				"ApplyCreatePatch",
+				"ApplyDeletePatch",
+				"ApplyUpdatePatch",
+				"GetCurrentState",
+				"GetDesiredState",
+				"NewUpdatePatch",
+				"ApplyCreatePatch",
+				"ApplyDeletePatch",
+				"ApplyUpdatePatch",
+				"GetCurrentState",
+				"GetDesiredState",
+				"NewUpdatePatch",
+				"ApplyCreatePatch",
+				"ApplyDeletePatch",
+				"ApplyUpdatePatch",
+			},
+		},
+	}
+
+	for i, tc := range testCases {
+		var err error
+
+		tr := &testCRUDResource{
+			Error:       fmt.Errorf("test error"),
+			ErrorCount:  tc.ErrorCount,
+			ErrorMethod: tc.ErrorMethod,
+		}
+
+		var r resource.Interface
+		{
+			c := crud.ResourceConfig{
+				CRUD:   tr,
+				Logger: microloggertest.New(),
+			}
+
+			r, err = crud.NewResource(c)
+			if err != nil {
+				t.Fatal(err)
+			}
+		}
+
+		rs := []resource.Interface{
+			r,
+		}
+		bf := func() backoff.Interface {
+			return backoff.NewMaxRetries(3, 0)
+		}
+
+		c := WrapConfig{
+			Logger:         microloggertest.New(),
+			BackOffFactory: bf,
+		}
+		wrapped, err := Wrap(rs, c)
+		if err != nil {
+			t.Fatal("test", i, "expected", nil, "got", err)
+		}
+
+		err = controller.ProcessUpdate(context.TODO(), nil, wrapped)
+		if err != nil {
+			t.Fatal("test", i, "expected", nil, "got", err)
+		}
+
+		if !reflect.DeepEqual(tc.ExpectedMethodOrder, tr.Order) {
+			t.Fatal("test", i, "expected", tc.ExpectedMethodOrder, "got", tr.Order)
+		}
+	}
+}
+
+// Test_Resource_Wrapper_RetryResource_ProcessUpdate_ResourceOrder ensures the
+// resource's methods are executed as expected when updating resources using the
+// wrapping retry resource.
+func Test_Resource_Wrapper_RetryResource_ProcessUpdate_ResourceOrder(t *testing.T) {
+	var err error
+
+	tr := &testCRUDResource{}
+
+	var r resource.Interface
+	{
+		c := crud.ResourceConfig{
+			CRUD:   tr,
+			Logger: microloggertest.New(),
+		}
+
+		r, err = crud.NewResource(c)
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	rs := []resource.Interface{
+		r,
+	}
+	bf := func() backoff.Interface {
+		return backoff.NewMaxRetries(0, 0)
+	}
+
+	c := WrapConfig{
+		Logger:         microloggertest.New(),
+		BackOffFactory: bf,
+	}
+	wrapped, err := Wrap(rs, c)
+	if err != nil {
+		t.Fatal("expected", nil, "got", err)
+	}
+
+	err = controller.ProcessUpdate(context.TODO(), nil, wrapped)
+	if err != nil {
+		t.Fatal("expected", nil, "got", err)
+	}
+
+	e := []string{
+		"GetCurrentState",
+		"GetDesiredState",
+		"NewUpdatePatch",
+		"ApplyCreatePatch",
+		"ApplyDeletePatch",
+		"ApplyUpdatePatch",
```

Comment:
Can we add a case with `ErrorCount: 0` to the `Test_Resource_Wrapper_RetryResource_ProcessDelete_ResourceOrder_RetryOnError` instead of having this test?

Reply:
I have so much no interest in making this "better". These tests were actually supposed to be re-activated by you and I asked for it several times. I want to get it merged as much as it is to be honest.

Reply:
We talked and agreed to go with what we have for now.